### PR TITLE
Claim trapped asset

### PR DESCRIPTION
**Pull request by yrong**
Resolves: https://linear.app/snowfork/issue/SNO-888

For comment https://linear.app/snowfork/issue/SNO-888#comment-4ff325b3 this is the prerequisite to implement ClaimAsset by original sender(no governance required).

Still require a companion PR on Gateway to 
- include the original sender into `sendToken`
- add another external call `claimTrappedToken`

### CODE REVIEW

File: cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs [hunk 1]
```diff
@@ -535,3 +537,87 @@ fn send_token_from_ethereum_to_asset_hub_fail_for_insufficient_fund() {
...
+				AccountKey20 { network: None, key: [1u8; 20] },
+			]),
+		};
+
+		// Check that asset trapped as original sender
+		assert_expected_events!(
+			AssetHubRococo,
+			vec![
+				RuntimeEvent::PolkadotXcm(pallet_xcm::Event::AssetsTrapped { origin,.. }) => {
+					origin: *origin == expected_origin,
```
Comment (alistair-singh) [inline cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs:618]: Nice! Does this test pass?

And if so we should extent the test to actually claim an asset.

This also maybe useful in the future: https://github.com/paritytech/xcm-format/blob/master/proposals/0037-custom-asset-claimer.md

Comment (yrong) [reply cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs:618]: Yeah, work as expected. Thanks for the sharing.

Comment (yrong) [reply cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs:618]: https://github.com/Snowfork/polkadot-sdk/pull/121/commits/1ae2af6ad03381c6fa501f6bf22a7cc08b6c300d

The `send_token_from_ethereum_to_asset_hub_trapped_and_then_claimed` test works as expected including the   claim flow.

File: bridges/snowbridge/primitives/router/src/inbound/mod.rs [hunk 2]
```diff
@@ -244,12 +278,12 @@ where
 		let inbound_queue_pallet_index = InboundQueuePalletInstance::get();
 
 		let mut instructions = vec![
-			ReceiveTeleportedAsset(total_fee_asset.into()),
-			BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited },
 			DescendOrigin(PalletInstance(inbound_queue_pallet_index).into()),
 			UniversalOrigin(GlobalConsensus(network)),
-			ReserveAssetDeposited(asset.clone().into()),
-			ClearOrigin,
```
Comment (yrong) [inline bridges/snowbridge/primitives/router/src/inbound/mod.rs:252]: The reason not use `Teleport` here is for the case when failed at executing the `BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited }` for insufficient fee, the token(weth) is in registry yet so can't be claimed and lost.

File: bridges/snowbridge/primitives/router/src/inbound/mod.rs [hunk 3]
```diff
@@ -264,17 +298,22 @@ where
...
 				]);
 			},
 			None => {
 				instructions.extend(vec![
-					// Deposit asset to beneficiary.
-					DepositAsset { assets: Definite(asset.into()), beneficiary },
+					// Deposit both asset and fees left to beneficiary so the fees will not get
+					// trapped. Another benefit is when fees left more than ED could be used to
+					// create the dest account in case it does not exist.
+					DepositAsset { assets: Wild(AllCounted(2u32)), beneficiary },
```
Comment (yrong) [inline bridges/snowbridge/primitives/router/src/inbound/mod.rs:316]: Benefit to use `Wild` Assets instead.

File: bridges/snowbridge/primitives/router/src/inbound/mod.rs [hunk 4]
```diff
@@ -244,12 +278,12 @@ where
...
-			BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited },
 			DescendOrigin(PalletInstance(inbound_queue_pallet_index).into()),
 			UniversalOrigin(GlobalConsensus(network)),
-			ReserveAssetDeposited(asset.clone().into()),
-			ClearOrigin,
+			// Todo: Hardcode for the POC and will pass original sender from Ethereum side
+			DescendOrigin([AccountKey20 { network: None, key: [1u8; 20] }].into()),
+			ReserveAssetDeposited(vec![total_fee_asset, asset.clone()].into()),
+			BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited },
 		];
```
Comment (yrong) [inline bridges/snowbridge/primitives/router/src/inbound/mod.rs:287]: Replace `ReceiveTeleportedAsset ` with `ReserveAssetDeposit` for xcm fees. 

The reason is for the former when failed at executing the `BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited }` for insufficient fee, the token(weth) is not in registry yet so can't be claimed thus lost.

Though it require xcm config on AssetHub `IsReserve` to set Ethereum as trusted reserve location for the RelayToken(i.e. `DOT`).

File: cumulus/parachains/runtimes/assets/asset-hub-rococo/src/xcm_config.rs [hunk 5]
```diff
@@ -559,6 +562,8 @@ impl xcm_executor::Config for XcmConfig {
 	type IsReserve = (
 		bridging::to_westend::IsTrustedBridgedReserveLocationForConcreteAsset,
 		bridging::to_ethereum::IsTrustedBridgedReserveLocationForForeignAsset,
+		bridging::to_ethereum::IsTrustedBridgedReserveLocationForFeeAsset,
+		bridging::to_ethereum::IsTrustedBridgedReserveLocationForFeeAssetFromChildSovereign,
```
Comment (yrong) [inline cumulus/parachains/runtimes/assets/asset-hub-rococo/src/xcm_config.rs:566]: Set Ethereum(with account as child sovereign) as the trusted reserve location for the FeeToken(i.e. DOT).

**Review by vgeddes — COMMENTED** [review]
I understand this PR solves the problem, in the Ethereum->AssetHub direction, of users losing assets on AssetHub if not enough fees are provided to buy execution weight.

There are also other locations where assets can be also be lost or trapped (like for example on BridgeHub, in the AssetHub->BridgeHub direction).

I observe that this PR makes a lot of changes to our existing protocol - new message types, new APIs, updated XCM scripts. To support only the Ethereum->AssetHub case, where assets trapped on AssetHub. 

I observe in Rococo (and presumably Polkadot) that on AssetHub XCM fees are very low, like 0.01 ROC (in dollars, less than 10 cents). With that in mind, if we configure our bridge to teleport double or even three times the necessary amount of fees, then we greatly minimise the chance of lost funds.

I know there are other ways that XCM messages can fail on AssetHub - for example, the destination account or parachain sovereign account not existing. However we can check for that in the frontend UI.

I think what I am trying to say is that for this specific case, there are some simple mitigations we can make to prevent loss of funds. They are not ideal, but they should work.

File: bridges/snowbridge/primitives/router/src/inbound/mod.rs [hunk 6]
```diff
@@ -56,6 +56,21 @@ pub enum Command {
...
+		sender: H160,
+		/// The destination for the transfer
+		destination: Destination,
+		/// Amount of token to claim
+		token_amount: u128,
+		/// Amount of fee to claim
+		fee_amount: u128,
+		/// XCM execution fee on AssetHub
+		asset_hub_fee: u128,
+	},
```
Comment (yrong) [inline bridges/snowbridge/primitives/router/src/inbound/mod.rs:73]: So the original sender can claim the token trapped on AH.

**Pull request closed without merging**